src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs

```diff
@@ -128,30 +148,77 @@ private static (bool CanMapWithLinq, string? CollectMethod) ResolveCollectMethod
+        if (SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.IEnumerableT))
             return (true, null);
 
         // if the target is IReadOnlyCollection<T>
         // and the count of the source is known (array, IReadOnlyCollection<T>, ICollection<T>) we collect to array
         // for performance/space reasons
-        var targetIsReadOnlyCollection = SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.IReadOnlyCollection);
+        var targetIsReadOnlyCollection = SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.IReadOnlyCollectionT);
         var sourceCountIsKnown =
             ctx.Source.IsArrayType()
-            || ctx.Source.ImplementsGeneric(ctx.Types.IReadOnlyCollection, out _)
-            || ctx.Source.ImplementsGeneric(ctx.Types.ICollection, out _);
+            || ctx.Source.ImplementsGeneric(ctx.Types.IReadOnlyCollectionT, out _)
+            || ctx.Source.ImplementsGeneric(ctx.Types.ICollectionT, out _);
         if (targetIsReadOnlyCollection && sourceCountIsKnown)
             return (true, ToArrayMethodName);
 
         // if target is a IReadOnlyCollection<T>, IList<T>, List<T> or ICollection<T> with ToList()
         return targetIsReadOnlyCollection
-            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.IReadOnlyList)
-            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.IList)
-            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.List)
-            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.ICollection)
+            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.IReadOnlyListT)
+            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.IListT)
+            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.ListT)
+            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.ICollectionT)
             ? (true, ToListMethodName)
             : (false, null);
     }
 
+    private static LinqEnumerableMapping? TryBuildImmutableLinqMapping(
+       MappingBuilderContext ctx,
+       ITypeMapping elementMapping)
+    {
+        var collectMethod = ResolveImmutableCollectMethod(ctx);
+        if (collectMethod is null)
+            return null;
+
+        var selectMethod = elementMapping.IsSynthetic
+            ? null
+            : ResolveLinqMethod(ctx, SelectMethodName);
+
+        return new LinqEnumerableMapping(ctx.Source, ctx.Target, elementMapping, selectMethod, collectMethod);
+    }
+
+    private static IMethodSymbol? ResolveImmutableCollectMethod(MappingBuilderContext ctx)
+    {
+        if (ctx.Target.ImplementsGeneric(ctx.Types.ImmutableArrayT, out _))
```

Comment:
First I was wondering why this works as `ImplementsGeneric` expects the `genericInterfaceSymbol` parameter to be an interface. It does actually work, since `ImplementsGeneric` does have a `SymbolEqualityComparer.Default.Equals(t.OriginalDefinition, genericInterfaceSymbol)` check at the beginning of the implementation.
I would consider this an implementation detail of the  `ImplementsGeneric` method and instead of calling this method replace these calls with `SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.ImmutableArrayT)`.